crates/core/tedge_actors/src/servers/builders.rs

```diff
@@ -204,6 +212,21 @@ impl<S: Server, K> ServiceProvider<S::Request, S::Response, NoConfig> for Server
     }
 }
 
+pub struct ReplyToRequester;
```

Comment:
No `NoConfig` cannot be used here.

The issue is that we want clients connect a server along two different ways (see https://github.com/thin-edge/thin-edge.io/pull/2600#discussion_r1500284821). Hence, the server actor builder must provides *two* different implementations of `ServiceProvider` and the client must be able to pick one or the other (as done by `FirmwareManagerBuilder::try_new()`.

That said I agree that something is not satisfying.

- I will check again why rust rejects as conflicting implementations for `ServiceProvider<Req,Res,NoConfig>` and   `ServiceProvider<Envelope<Req,Res>, NoMessage,NoConfig>`.
- I will try to find a way to hide this issue inside the actor crate, so a client has only to peek a client message box type.
- Going deeper, it a good time to revisit the `ServiceProvider` & `ServiceConsumer` traits vs `MessageSource` and `MessageSink`. Indeed, a server is now a `MessageSink<Envelope<Req,Res>>`.

Reply:
So, I finally removed the need for `ReplyToRequester`. See https://github.com/thin-edge/thin-edge.io/pull/2600/commits/9dd8ff64d2d0af1da23dd8d7a48ad2ac8000061e and https://github.com/thin-edge/thin-edge.io/pull/2600/commits/363af5dfa9058960d04a368d90f1bc006ddf0d00

The idea is that `ServerBoxBuilder` implements `MessageSink<Envelop<Req,Res>, NoConfig>` rather than `ServiceProvider<Envelop<Req,Res>, NoMessage, ReplyToRequester>`.

I also introduced a trait alias `ServerEndpoint<Req,Res>` for `MessageSink<Envelop<Req,Res>, NoConfig>`, to make clearer the code connecting a client to a server. See for instance `FirmwareManagerBuilder::try_new()`. Note that the client decide how a server is connected using a specific trait for the server. A `ServerEndpoint`, returns the response to the requester (as for download requests), while a `ServiceProvider` returns the responses on a different channel (as for timer requests).

I'm not fully happy with the name `ServerEndpoint`, though.

Reply:
> I'm not fully happy with the name `ServerEndpoint`, though.

I propose `Service` in https://github.com/thin-edge/thin-edge.io/pull/2600/commits/fa18fb4b7a900b1305b014f768ddc96df0520691